qiskit/algorithms/optimizers/cmaes.py

```diff
@@ -0,0 +1,400 @@
+        of the distribution.
+    """
+
+    x_fun_translation: Optional[Union[POINT, List[POINT]]] = None
+
+
+@dataclass
+class CMAESState(OptimizerState):
+    """
+    Args:
+        p_sigma: Evolution path for the standard deviation of the distribution.
+        p_c: Evolution path for the covariance matrix.
+        cov_matrix: Covariance matrix.
+        eigenvectors:   Eigenvectors of the covariance matrix.
+        std_vector: Standard deviation in each principal axis of the covariance matrix.
+        They can also be tought as the square root of the eigenvalues of the covariance matrix.
+        sigma: The standard deviation of the distribution.
+    """
+
+    # x will be treated as the mean
+    p_sigma: POINT
+    p_c: POINT
+    cov_matrix: np.ndarray  # pylint: disable=invalid-name
+    eigenvectors: np.ndarray
+    std_vector: POINT  # Will store the sqrt of the diagonal elements
+    sigma: float
+
+    # Parameters for the stoping criteria
+    _best_func_hist: List[float]
+    _median_func_hist: List[float]
+    _worst_func_eval: Optional[float]
+
+
+class CMAES(SteppableOptimizer):
+    """
+    Covariance Matrix Adaptation Evolution Strategy minimization routine.
+    This Optimizer is based on the paper: `arXiv:1604.00772 <https://arxiv.org/abs/1604.00772>`_.
+    CMA-ES is an evolution strategy based on sampling points from a multivariate normal distribution
+    with a given centroid and covariance matrix.
+    From the :math:`\lambda\ points sampled in generation ``g``, a new centroid is calculated as:
+    .. math::
+        x^{g+1} = \sum_{i=1}^{\mu} w_i x_{i:\lambda}^{g}
+        where :math:`\sum_{i=1}^{\mu} w_i = 1` and :math:`x_{i:\lambda}^{g}` are the best :math:`\mu`
+        points from generation ``g`` ordered in function of their objective function image.
+
+    A new covariance matrix is also computed taking into account not only the current population,
+    but also the previous covariance matrix.
+
+    The optimization is stopped either at the begining of the optimization process if the initial
+    parameters were not well chosen, or once the objective function stops decreasing for a given
+    number of generations. There are several criteria that can be manually set by the user, and
+    are well described in the paper.
+
+    One evaluation with a funciton evaluation outlier with a significantly low value is not
+    enough to stop the optimization process or significantly disturb the position of the centroid,
+    and an outlier with a significantly high value would simply be ignored by the algorithm.
+    Therefore, this optimizer is resistant to noise in the objective function and can be used with
+    non-smooth functions.
+    """
+
+    def __init__(
+        self,
+        maxiter: int = 100,
+        callback: Optional[CALLBACK] = None,
+        sigma_0: float = 0.5,
+        weights: Optional[np.ndarray] = None,
+        termination_criteria: Optional[Dict[str, Optional[float]]] = None,
```

Comment:
For the termination criteria, I am looking at the value which says Optional[float] - is there some good reason to have say to have a key where it has a value of None versus just omitting the key?

Reply:
Yes, the termination criteria inside of the dictionary are the ones that will be used. The float value on each one of the keys is the tolerance for that particular termination criteria. It might be the case that the user wants to use a termination criteria but doesn't know what is the best tolerance to use. In that case he can leave it to None and a default value will be used for that termination criteria.